lib/gateway.js

```diff
@@ -0,0 +1,250 @@
+          method: 'POST',
+          body: JSON.stringify({
+            query: `
+            query ServiceInfo {
+              _service {
+                sdl
+              }
+            }
+            `
+          })
+        })
+
+        const schemaDefinition = response.json.data._service.sdl
+
+        const typeMap = createTypeMap(schemaDefinition)
+
+        return {
+          schemaDefinition,
+          typeMap
+        }
+      }
+    }
+
+    serviceMap[service.name] = serviceConfig
+  }
+
+  await Promise.all(services.map(service => {
+    return serviceMap[service.name].init().then(({ schemaDefinition, typeMap }) => {
+      serviceMap[service.name].schemaDefinition = schemaDefinition
+      serviceMap[service.name].typeMap = typeMap
+      serviceMap[service.name].types = new Set(Object.keys(typeMap))
+    })
+  }))
+
+  return serviceMap
+}
+
+function isDefaultType (type) {
+  return [
+    '__Schema',
+    '__Type',
+    '__Field',
+    '__InputValue',
+    '__EnumValue',
+    '__Directive'
+  ].includes(type)
+}
+
+function makeResolveQuery (service) {
+  return function (root, args, context, info) {
+    const { fieldName, returnType, operation } = info
+
+    // Must deep-copy the operation because if we modify it (remove an item from the selection set)
+    // then the resolvers for those fields won't be executed
+    const currentOperation = JSON.parse(JSON.stringify(operation))
+    currentOperation.selectionSet.selections[0].selectionSet.selections = currentOperation.selectionSet.selections[0].selectionSet.selections.filter(selection => service.typeMap[returnType].has(selection.name.value))
+
+    return service.request({
+      method: 'POST',
+      body: JSON.stringify({
+        query: print(currentOperation)
+      })
+    }).then(response => {
+      // TODO handle GraphQL error from remote call
+      return {
+        ...response.json.data[fieldName]
+      }
+    })
+  }
+}
+
+function makeFieldResolver (service) {
+  return function (root, args, context, info) {
+    // TODO implement field resolver
+
+    return null
+  }
+}
+
+function defineResolvers (schema, typeToServiceMap, serviceMap) {
+  const types = schema.getTypeMap()
+
+  for (const type of Object.values(types)) {
+    if (isObjectType(type) && !isDefaultType(type.name)) {
+      const serviceForType = typeToServiceMap[type.name]
+
+      for (const field of Object.values(type.getFields())) {
+        const fieldType = getNamedType(field.type)
+        if (!isScalarType(fieldType)) {
+          const serviceForFieldType = typeToServiceMap[fieldType]
+
+          if (serviceForFieldType !== serviceForType) {
+            if (serviceForType === null) {
+              field.resolve = makeResolveQuery(serviceMap[serviceForFieldType])
+            } else {
+              field.resolve = makeFieldResolver(serviceMap[serviceForFieldType])
+            }
+          }
+        }
+      }
+    }
+  }
+}
```

Comment:
Would you mind to add a bit of comments to this function, it's very complex.

Reply:
Added comment, will do the same for the `makeResolver` method as that one is complex as well.